File: swift/1345_JumpGameIV.swift
class Solution {
    func minJumps(_ arr: [Int]) -> Int {
        var N = arr.count
        var graph = [Int:[Int]]()
        for (i, n) in arr.enumerated() {
            if graph[n] != nil {
                graph[n]!.append(i)
            } else {
                graph[n] = [i]
            }
        }
        
        var count = 0
        var visited = [Bool](repeating: false, count: N)
        var nodes = [0]
        
        while !nodes.isEmpty {
            var nextNodes = [Int]()
            for node in nodes {
                if node == N - 1 {
                    return count
                }
                
                for nei in graph[arr[node]]! {
                    if !visited[nei] {
                        nextNodes.append(nei)
                        visited[nei] = true
                    }
                }
                
                for nei in [node - 1, node + 1] {
                    if nei >= 0 && nei < N && !visited[nei] {
                        nextNodes.append(nei)
                        visited[nei] = true
                    }
                }
                
                graph[node] = []
            }
            count += 1
            nodes = nextNodes
        }
        return -1
    }
}

File: java/1345_JumpGameIV.java
class Solution {
    public int minJumps(int[] arr) {
        HashMap<Integer, List<Integer>> graph = new HashMap<Integer, List<Integer>>();
        for (int i = 0; i < arr.length; ++i) {
            List<Integer> tmp;
            if (graph.get(arr[i]) == null) {
                tmp = new ArrayList<Integer>();
            } else {
                tmp = graph.get(arr[i]);
            }
            tmp.add(i);
            graph.put(arr[i], tmp);
        }
        
        int count = 0;
        List<Integer> nodes = new ArrayList<Integer>() {{ add(0); }};
        boolean[] visited = new boolean[arr.length];
        while (!nodes.isEmpty()) {
            List<Integer> nextNodes = new ArrayList<Integer>();
            for (int node : nodes) {
                if (node == arr.length - 1) return count;

                for (int nei : graph.get(arr[node])) {
                    if (!visited[nei]) {
                        visited[nei] = true;
                        nextNodes.add(nei);
                    }
                }
                
                for (int nei : new int[]{node - 1, node + 1}) {
                    if (nei >= 0 && nei < arr.length && !visited[nei]) {
                        visited[nei] = true;
                        nextNodes.add(nei);
                    }
                }
                
                graph.put(arr[node], new ArrayList<Integer>());
            }
            
            nodes = nextNodes;
            count++;
        }
        return -1;
    }
}
